builtin/logical/database/path_roles.go

```diff
@@ -2,263 +2,606 @@ package database
+}
+
+func (b *databaseBackend) pathRoleList(ctx context.Context, req *logical.Request, data *framework.FieldData) (*logical.Response, error) {
+	path := databaseRolePath
+	if req.Path == "static-roles/" {
+		path = databaseStaticRolePath
+	}
+	entries, err := req.Storage.List(ctx, path)
+	if err != nil {
+		return nil, err
+	}
+
+	return logical.ListResponse(entries), nil
+}
+
+func (b *databaseBackend) pathRoleCreateUpdate(ctx context.Context, req *logical.Request, data *framework.FieldData) (*logical.Response, error) {
+	name := data.Get("name").(string)
+	if name == "" {
+		return logical.ErrorResponse("empty role name attribute given"), nil
+	}
+
+	role, err := b.Role(ctx, req.Storage, data.Get("name").(string))
+	if err != nil {
+		return nil, err
+	}
+	if role == nil {
+		role = &roleEntry{}
+	}
+
+	if err := pathRoleCreateUpdateCommon(ctx, role, req.Operation, data); err != nil {
+		return logical.ErrorResponse(err.Error()), nil
+	}
+
+	// TTLs
+	{
+		if defaultTTLRaw, ok := data.GetOk("default_ttl"); ok {
+			role.DefaultTTL = time.Duration(defaultTTLRaw.(int)) * time.Second
+		} else if req.Operation == logical.CreateOperation {
+			role.DefaultTTL = time.Duration(data.Get("default_ttl").(int)) * time.Second
 		}
-		if len(role.Statements.Renewal) == 0 {
-			data["renew_statements"] = []string{}
+		if maxTTLRaw, ok := data.GetOk("max_ttl"); ok {
+			role.MaxTTL = time.Duration(maxTTLRaw.(int)) * time.Second
+		} else if req.Operation == logical.CreateOperation {
+			role.MaxTTL = time.Duration(data.Get("max_ttl").(int)) * time.Second
 		}
+	}
+
+	// Store it
+	entry, err := logical.StorageEntryJSON(databaseRolePath+name, role)
+	if err != nil {
+		return nil, err
+	}
 
-		return &logical.Response{
-			Data: data,
-		}, nil
+	if err := req.Storage.Put(ctx, entry); err != nil {
+		return nil, err
 	}
+
+	return nil, nil
 }
 
-func (b *databaseBackend) pathRoleList() framework.OperationFunc {
-	return func(ctx context.Context, req *logical.Request, data *framework.FieldData) (*logical.Response, error) {
-		entries, err := req.Storage.List(ctx, "role/")
-		if err != nil {
-			return nil, err
+func (b *databaseBackend) pathStaticRoleCreateUpdate(ctx context.Context, req *logical.Request, data *framework.FieldData) (*logical.Response, error) {
+	name := data.Get("name").(string)
+	if name == "" {
+		return logical.ErrorResponse("empty role name attribute given"), nil
+	}
+
+	role, err := b.StaticRole(ctx, req.Storage, data.Get("name").(string))
+	if err != nil {
+		return nil, err
+	}
+
+	if role == nil {
+		role = &roleEntry{}
+	}
+
+	if err := pathRoleCreateUpdateCommon(ctx, role, req.Operation, data); err != nil {
+		return logical.ErrorResponse(err.Error()), nil
+	}
+
+	// createRole is a boolean to indicate if this is a new role creation. This
+	// is used to ensure we do not allow an existing role to be "migrated" to
+	// role with a static account. If createRole is false and static_account
+	// data is given, return an error
```

Comment:
Is this still true now that we have separate storage paths? I don't see how a non-static role could be updated to a static role.

Reply:
Correct, this comment is out of date. I updated it in 21edf957c. `createRole` is still used later down the codepath